graph-subscriptions-rs/src/lib.rs

```diff
@@ -60,108 +61,101 @@ pub struct TicketPayload {
-        Ok((payload, *signature))
+        Ok((payload, signature))
     }
 
     pub fn to_ticket_base64(
         &self,
-        domain_separator: &DomainSeparator,
-        key: &PrivateKey,
+        domain: &TicketVerificationDomain,
+        wallet: &Wallet<SigningKey>,
     ) -> anyhow::Result<String> {
-        let ticket = self.encode(domain_separator, key)?;
+        let ticket = self.encode(domain, wallet)?;
         Ok(BASE64_URL_SAFE_NO_PAD.encode(ticket))
     }
 
-    pub fn verify(
-        &self,
-        domain_separator: &DomainSeparator,
-        signature: &[u8; 65],
-    ) -> anyhow::Result<Address> {
-        let hash = eip712::sign_hash(domain_separator, self);
-        let signature = Signature {
-            r: signature[0..32].into(),
-            s: signature[32..64].into(),
-            v: signature[64].into(),
-        };
-        let recovered_signer = signature.recover(RecoveryMessage::Hash(hash.into()))?;
-        ensure!(&recovered_signer == &self.signer);
-        Ok(self.signer)
-    }
-
     pub fn encode(
         &self,
-        domain_separator: &DomainSeparator,
-        key: &PrivateKey,
+        domain: &TicketVerificationDomain,
+        wallet: &Wallet<SigningKey>,
     ) -> anyhow::Result<Vec<u8>> {
-        let (sig, r) = self.sign_hash(domain_separator, key)?;
         let mut buf = serde_cbor_2::ser::to_vec(self)?;
-        buf.append(&mut sig.into());
-        buf.push(r);
+        buf.append(&mut self.sign_hash(domain, wallet)?.to_vec());
         Ok(buf)
     }
 
     pub fn sign_hash(
         &self,
-        domain_separator: &DomainSeparator,
-        key: &PrivateKey,
-    ) -> anyhow::Result<([u8; 64], u8)> {
-        Ok(eip712::sign_typed(domain_separator, self, key)?)
+        domain: &TicketVerificationDomain,
+        wallet: &Wallet<SigningKey>,
+    ) -> anyhow::Result<Signature> {
+        let hash = hash_message(self.verification_message(domain));
+        Ok(wallet.sign_hash(hash)?)
     }
 
-    pub fn eip712_domain(chain_id: u64, contract_address: Address) -> Eip712Domain {
-        let mut chain_id_bytes = [0_u8; 32];
-        chain_id_bytes[24..].clone_from_slice(&chain_id.to_be_bytes());
-        Eip712Domain {
-            name: "Graph Subscriptions".to_string(),
-            version: "0".to_string(),
-            chain_id: eip712::U256(chain_id_bytes),
-            verifying_contract: eip712::Address(contract_address.0),
-            salt: [42_u8; 32],
+    pub fn verify(
+        &self,
+        domain: &TicketVerificationDomain,
+        signature: &Signature,
+    ) -> anyhow::Result<Address> {
+        let hash = hash_message(self.verification_message(domain));
+        let recovered_signer = signature.recover(hash)?;
+        ensure!(
+            recovered_signer == self.signer,
+            "recovered signer does not match claim"
+        );
+        Ok(self.signer)
+    }
+
+    pub fn verification_message(&self, domain: &TicketVerificationDomain) -> String {
```

Comment:
Remember also that ticket also needs to include the Gateway address. This is to prevent, among other things, attacks where a Gateway proxies requests.

Reply:
True. I forgot about that. We may need to document this constraint in the future.